File: algorithm/96. Unique Binary Search Trees/96.py
class Solution(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp=[1]*(n+1)
        for i in range(2,n+1):
        	dp[i]=0
        	for j in range(1,i+1):
        		dp[i]+=dp[j-1]*dp[i-j]
        return dp[n]

File: algorithm/96. Unique Binary Search Trees/96.java
class Solution {
    public int numTrees(int n) {
        int[] F = new int[n+1];
        F[0] = 1; F[1] = 1;
        for(int i = 2; i <= n; i++) {
            for(int j = 0; j <= i -1; j++) {
                F[i] += F[j] * F[i - 1 -j];
            }
        }
        return F[n];
    }
}
